### DISCUSSION TITLE
Welcome to Smart_Powermeter Discussions!

### DISCUSSION THREAD

**Discussion by JGAguado** [JGAguado/Smart_Powermeter#18] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about and can't find on the [documentation FAQs](https://smart-powermeter.readthedocs.io/en/latest/faq.html).
  * Share ideas.
  * Engage with other community members.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by CNC-Maniac** [comment]
Hi, I bought your smartmeter board an I love the clean design.
Do you have another compiled firmware so that there is no need to have HA and VSC installed?
I am using IoBroker with ESPHOme and I'm not able to get this thing to run.
It would be perfect to have a web-based configuration, so that you just choose the right connection and set the parameters for calibration andMQTT.
VG
Axel

**Reply by JGAguado** [reply]
Hi Axel, thanks for your support!
With ESPHome you have support for a web-server component (https://esphome.io/components/web_server). I've never used it, but should give you access to interface with other components, like a number component (https://esphome.io/components/number/index.html) to later use as calibration coefficients. 
Regarding the MQTT, it has also support (https://esphome.io/components/mqtt). Again, I haven't used it cause I normally use the native API connection with HA, sorry 😐.
Did you get to deploy an ESPHome setup? I mean even without using HA, like described on the documentation (https://smart-powermeter.readthedocs.io/en/latest/programming.html#flash-tools).
LG,
Jon

**Reply by CNC-Maniac** [reply]
Hi Jon, I flashed the bin like in the docs and now I can see that also the ePaper display seems to work. Anyway It would realy helpful to have a version with at least a web-config, like espeasy or tosmata. Otherwise it is not usable without HA installed.
I would like to order more devices, but therefore I need the things above.
LG
Axel

**Reply by JGAguado** [reply]
Let me see what I can do 😉

**Reply by JGAguado** [reply]
I created a modified version of my standard config.yaml file, added the webserver, the mqtt component and set some input variables (configurable through the webserver) for calibrating each of the probes in the form $a*x+b$ where a and b are linear coefficients that you can calculate with a simple regression model based on your actual needs, and the $x$ is just the raw value.

You can directly copy the new [config.yaml](https://github.com/JGAguado/Smart_Powermeter/blob/main/docs/source/files/configuration_1.yaml) and remember to add the[ www.js](https://github.com/JGAguado/Smart_Powermeter/blob/main/docs/source/files/www.js) file to the same path as well, or any similiar, this is just an example.

The webserver looks like this:
![image](https://github.com/JGAguado/Smart_Powermeter/assets/7847822/32097157-2606-4fcc-9df4-57487c643f5f)

Not too elegant, but should do the job if you don't want to be linked to HA.

**Reply by CNC-Maniac** [reply]
Super, thats a big improvement for me. I had to remove all the stuf for the display, because I still have the error message:
` Please update your pillow installation to 10.2.0. (pip install "pillow==10.2.0").`
But this seems to be a ESPHome problem. I also didn't use the MQTT part yet, because ESPHome is running inside the IoBroker with the adapter and so I get the data without MQTT anyway. Tomorrow I will test the measurements with the web-server and my equipment. Than I can order the next devices :-) Do you know when the upgraded version with voltage measurement will be available? 
VG Axel

**Reply by JGAguado** [reply]
The Pillow thingy happened to me a lot when compiling it from the CLI. Let me know if it would help you if I compile it. (you will have to add it to your wifi network, but shouldn't be an issue as it creates a hotspot after 2 mins without connection).

For the voltage measurement we will have to wait probably until the end of the summer 😉

**Reply by CNC-Maniac** [reply]
That sounds good, but how can I connect it to my ESPHome inside IoBroker? I mean that in AP mode, it will ask for the right wlan for connecting. After this it should be visible via web browser. But how to connect than to ESPHome again? Will it be recognized be ESPHome?
Maybe I'll create something for the voltage measurement on EasyEDA.

**Reply by CNC-Maniac** [reply]
To more things: 1 with the code above, the states doesn't move. So I went back to the old code, add the web-server and everything seems to be okay so far. 2 Is it possible to read negative values?

**Reply by JGAguado** [reply]
In the code I recently added:

0. The broker is a Mosquitto MQTT running on the same Home Assistant Raspberry as an Add-on. So on the ESPHome config you just need to add the broker's IP, and username  + password so it streams the data there.
1. In my case all the states move, except for the ones containing the coefficients a & b (which are user editable). The thing is that if your coefficient $a$ is very small, the output will barelly move. Remember that the coefficients are just multiplying the raw input in the way $a*x + b$, where $x$ is the equation.
2. No, only positive.

**Comment by n2kbg** [comment]
I made two modes to my board:

1) Added line voltage sensing to one input. Use an AC transformer to isolate line voltage from the board and reduce it to a safe level (I used an old 120v/9vac "wall wart" I had from some old radio or something.) Reduce the output of that transformer with two resistors configured as a voltage divider so the peak of the AC waveform (1.4 * volts RMS) will be within the biased range of the inputs.  Feed that conditioned signal into a port after removing the 22 ohm burden resistor (which is not needed for voltage-based inputs). Now you can calculate the instantaneous power more accurately by multiplying the actual line voltage instead of using a fixed voltage in the calculation. 

The ESPHome "ct_clamp" platform will sample the input and calculate the RMS voltage just like it does when using it for current measurement. You need to add "device_class: voltage" and "unit_of_measurement: V" to the esphome entry for the probe that you are using to measure voltage. You can edit the setting for that probe in the display too (I changed the "A" to a "V" and adjusted the min/max to be 112/128 which is just outside the 5% that it "should" be in the USA.

It should also be possible to calculate the power factor now too since both the current and voltage waveforms are being sampled, but in the US residential customers aren't billed for reactive power so I'm not going to put in the effort of figuring that out in ESPHome.

2) The ESP32-S2 ADC reads 0-2.5v. As designed, the board biases the inputs to 1.625v which is above the midpoint of the ADC input. With 100A/50ma current transformers and the 22 ohm burden, you will go above the maximum adc input voltage at the maximum current level and get incorrect readings. Replacing one of the 10k ohm resistors in the voltage dividers (R3,6,9,12,15) with 15k ohm parts moves the bias down to ~1.3 volts giving enough headroom for those sensors. Amazon sells packs of assorted SMD resistors for under $10 USD, but removing and replacing the 0603 resistors by hand is left as an exercise to the reader... (Thank you for making the 22 ohm ones larger!)

**Reply by JGAguado** [reply]
Amazing @n2kbg! Very cool mods 😃I'm welcoming always any ideas to improve it!
1. I'm currently working on a "Pro" version for simultaneously monitoring voltage in a very similar config as you did 😄! I expect to release it around summer, but my time is limited and I'm going slower than expected ...
2. That's right. I didn't realized earlier because at home I'm far away from reaching 100A (~2.72V on the ADC input) on just one line, so I never had the issue of reading over the 2.5V limitation on the ADC (with the 11dB attenuation setting). I will open an issue and replace those resistors in the next version (V2R2) that will be launched on the next batch Together with a replacement of the SMD jacks with THT ones (more robust join). 
I appreciate the feedback on the 1206 package. However, I feel it would be more useful if, instead of having to desolder it, it would be better to remove the connection for the burden resistor through a cuttable jumper pad. Therefore I'll be adding this feature on V2R2 and replacing the 1206 with a 0603 (for keeping a compact format).

Thanks for the feedback! 

All best!

Jon

**Reply by n2kbg** [reply]
I would also like to see some change with the TRRS jacks so that it isn't necessary to pull the plug out 1mm. I'm not sure if different CT clamps have different wiring you need to support. I only have the 100A/50ma ones. Maybe make the traces cuttable to disconnect the unneeded contacts? I replaced the TRS plugs on my ct clamps with TRRS and only connected T & S. I would have cut or disconnected the pins on the jacks, but that isn't especially easy to do in the current design.

Another possible issue with the 100A/50ma CT clamps is they could have a "high" voltage on them (higher than the ESP32 would like) if they are sensing a conductor before they are connected to the board. Designing the jack so that it shorts the plug on the way in would be a useful addition if it is possible to do. You have RRT connected but maybe connecting SRR instead could accomplish this and avoid the 1mm pullout?

I didn't check the original wiring of the sensors before I cut the original plugs off so maybe none of this is possible.

**Reply by JGAguado** [reply]
Yes, on the new revision, together with new THT connectors (that should remove the need to keep the 1mm gap), only T&S will be connected, reducing the chance of this "high" voltage terminals to enter into a short when plugging the jack. Until now, I've seen that most of the CT probes are using the T&S, but I don't know if it is a standard.

**Reply by CNC-Maniac** [reply]
Maybe it would be good to have jumper to select the current range. For the most people in Europe, the config you use rigth now is ok, because normal houses are limited to max. 63A

**Comment by 2rkoester** [comment]
Got a question, just received my board last week and have it programmed and reading data in HAS. I read that this board also contains a bmp280 sensor so I have edited the .ymal file to enable i2c and the sensor, but the sensor fails to load. Is it possible it’s no longer part of what’s being shipped now?

**Reply by JGAguado** [reply]
Dear @2rkoester,

The BMP280 was integrated in one of the first prototypes of the V2, but since most of people let me know it was highly unnecessary, I took the decision of never integrating it on the boards that I started selling. However, I let open an I2C port in case anyone wants to integrate their own external sensors, but if you are just interested on the temperature you could access to the ESP32-S2 temperature sensor.
I'm sorry if the Hackaday post mislead you, I'll edit it to prevent further confusions  😉

**Comment by 2rkoester** [comment]
Thank you for the prompt response.Rodger KoesterSent from my iPhoneOn May 30, 2024, at 2:43 AM, J.G.Aguado ***@***.***> wrote:﻿
Dear @2rkoester,
The BMP280 was integrated in one of the first prototypes of the V2, but since most of people let me know it was highly unnecessary, I took the decision of never integrating it on the boards that I started selling. However, I let open an I2C port in case anyone wants to integrate their own external sensors, but if you are just interested on the temperature you could access to the ESP32-S2 temperature sensor.
I'm sorry if the Hackaday post mislead you, I'll edit it to prevent further confusions  😉
